### ISSUE TITLE
Multiple redundant invocations of the method com.vaadin.flow.router.HasUrlParameter#setParameter in Vaadin 24.4.9

### ISSUE THREAD

**Issue by alexanoid** [vaadin/flow#19822]
### Description of the bug

I have an AdministrationView:

```
@Route(value = "administration", layout = AdministrationLayout.class)
@PermitAll
public class AdministrationView extends BaseView implements BeforeEnterObserver {

    @Override
    public void beforeEnter(BeforeEnterEvent event) {

        removeAll();

        event.forwardTo(AdministrationRecruitersView.class);
    }

}
```

Available at the following URL:

http://localhost:8080/administration

As you can see, in the `beforeEnter `method, there is a forward to `AdministrationRecruitersView.class`.

The `AdministrationRecruitersView.class` itself is implemented as:

```
@Route(value = "administration/recruiters", layout = AdministrationLayout.class)
@PermitAll
public class AdministrationRecruitersView extends CatalogContainer<DecisionMatrix> implements HasUrlParameter<String> {

    @Override
    public void setParameter(BeforeEvent event, @OptionalParameter String jobIdParameter) {
    }

}
```

So, when navigating to AdministrationView (by http://localhost:8080/administration) and subsequently auto forwarding to AdministrationRecruitersView, the method AdministrationRecruitersView.setParameter() is called **3 times in a row**. If I navigate directly to AdministrationRecruitersView, i.e., to http://localhost:8080/administration/recruiters, the method AdministrationRecruitersView.setParameter() is correctly called only once, as it should be.

This issue started appearing in version 24.4.9, while in version 24.4.0.beta5 everything works correctly. I haven't changed anything in my code. 

Also, in version 24.4.9, the bug from https://github.com/vaadin/flow/issues/18736 has reappeared, but it’s not present in version 24.4.0.beta5, where everything works fine.

### Expected behavior

no redundant invocation of com.vaadin.flow.router.HasUrlParameter#setParameter

### Minimal reproducible example

n/a

### Versions

- Vaadin / Flow version: 24.4.9
- Java version: 17

**Comment by alexanoid** [comment]
Sorry to bother you, but is there a chance this will be fixed anytime soon?

**Comment by tepi** [comment]
Seems this is another issue with the `react-router` performing the full navigation cycle on server side when forwarding is used, and after that doing a `ui-navigate` event callback to the server which triggers the `setParameter` for a second time. That said, in my testing I only got two calls to `setParameter`, not three.

We are looking into these problems with `react-router`. In the meantime, if it is an option for you, please try setting system parameter `vaadin.react.enable=false` to revert back to using vaadin router. That seems to work correctly and produce only one of each calls as it should.

**Comment by alexanoid** [comment]
@tepi Hi! Could you please tell me in which version this fix will be released?

**Comment by tepi** [comment]
> @tepi Hi! Could you please tell me in which version this fix will be released?

Hi! The fix is released in 24.5.0.beta2 and will be released in Flow 24.4.9 which should be released within a week as part of Vaadin 24.4.13.

**Comment by alexanoid** [comment]
@tepi Thank you! Sorry for bothering you with additional questions – but do I need to change anything in the code to switch from version 24.4.* to 24.5.* or should it go smoothly?

**Comment by tepi** [comment]
There should not be breaking changes between those versions, version update is enough.

**Comment by alexanoid** [comment]
Hi @tepi 

I just upgraded to Vaadin 24.4.13, and unfortunately, the issue with the repeated triggering of `com.vaadin.flow.router.HasUrlParameter#setParameter#setParameter `is still present. Instead of 3 times, the number of calls has been reduced to 2. Something is still not right.

**Comment by tepi** [comment]
> Hi @tepi
> 
> I just upgraded to Vaadin 24.4.13, and unfortunately, the issue with the repeated triggering of `com.vaadin.flow.router.HasUrlParameter#setParameter#setParameter `is still present. Instead of 3 times, the number of calls has been reduced to 2. Something is still not right.

Could you provide some more context on how to reproduce the issue? Using Vaadin 24.4.13 and the original example from this ticket, as well as [the test built for this issue](https://github.com/vaadin/flow/pull/19959/files#diff-74f83b769312358581082dc0fe974e82a51c2bc6b82713051e0d474ca191ad4dR44-R60), there is only one call. So I suspect that there is something in your case that has not been revealed in this ticket yet.

**Comment by alexanoid** [comment]
@tepi I think I've localized the problem. In my previous example (with which I created this ticket), everything remains the same. If I simply go to localhost:8080/administration, everything works correctly, even with forwarding—`setParameter `is only called once.

But I also have `public class MainLayout extends AppLayout`, and in it, I create `Tabs `in the `Drawer`, where I add, for example:

```
RouterLink link = new RouterLink("Test Link", AdministrationView.class);
tabs.add(new Tab(link));
addToDrawer(tabs);
```

And now, when I click on this RouterLink, I get two calls to the `setParameter `method.

I hope this helps you localize the issue.

**Comment by tepi** [comment]
Confirmed. Judging from browser console, all seems to be fine in the uidl response: a `vaadin-navigate` event is dispatched with target `first` and callback `false`. Yet after this, _something_ anyways makes the callback to server, causing the second round of `beforeEnter`/`setParameter` calls. I'll investigate this sometime next week.

**Comment by alexanoid** [comment]
@tepi Thank you very much for correcting this! Could you please let me know in which release version this fix is expected?

**Comment by tepi** [comment]
Thanks to @mcollovati for the latest fix. This will be released in next 24.4 and 24.5 branch releases; Vaadin 24.4.14 and the first RC for Vaadin 24.5. Both should happen within a week.
